Added support for ports

evangelion1204 opened the pull request:
Closes #127

aik099 commented on library/QATools/QATools/PageObject/Url/Builder.php:
```diff
@@ -20,6 +20,9 @@
  */
 class Builder implements IBuilder
 {
+	const DEFAULT_PORT_HTTP = '80';
```
Any reason for storing numbers as string?

evangelion1204 replied:
Ok, thats true, in this case I will remove them.

aik099 commented on tests/QATools/QATools/PageObject/Url/BuilderTest.php:
```diff
@@ -140,6 +142,36 @@ public function buildDataProvider()
+					'scheme' => 'http',
+					'host' => 'domain.tld',
+					'path' => '/path',
+					'port' => '8080',
```
Any reason for specifying port as string?

evangelion1204 replied:
No particular, its true that they should be a number, I will change it.

evangelion1204 commented:
Ok, corrected.

evangelion1204 commented:
And the other stuff :smile: sorry.

aik099 commented on library/QATools/QATools/PageObject/Url/Builder.php:
```diff
@@ -104,6 +112,26 @@ public function build(array $params = array())
+
+		$default_ports = array('http' => 80, 'https' => 443);
+
+		if ( $default_ports[$this->getProtocol()] == $this->getPort() ) {
```
No Yoda Conditions please. The constant part should be on the right of comparison.

evangelion1204 replied:
Corrected and pushed.

aik099 commented:
Today’s review completed.

evangelion1204 commented:
Pushed.

aik099 commented:
Totally forgot that we have docs for annotations, so please describe port usage there as well.

aik099 commented on library/QATools/QATools/PageObject/Url/Builder.php:
```diff
@@ -104,6 +112,26 @@ public function build(array $params = array())
+
+		$default_ports = array('http' => 80, 'https' => 443);
+
+		if ( $this->getPort() == $default_ports[$this->getProtocol()] ) {
```
Use `===` here.

evangelion1204 commented:
I need to add the support in the annotations as well.

evangelion1204 commented:
Ok, corrected `===` and added a doc entry. But for all use cases that come to my mind I am not sure if we should add `port` to `@page-url` if somebody uses a custom port and secure he might need to switch the port.
* 8080 => http
* 8081 => https

aik099 commented:
Since the port is empty by default, then no matter if secure is used or not or it's value we're not adding port to url. If developer specifies both secure & port, then he should know that given port would be used regardless the protocol.

aik099 commented on docs/annotations.rst:
```diff
@@ -227,6 +227,16 @@ By specifying ``secure`` annotation parameter (3rd) value it's possible to force
     @page-url('url' => 'https://www.example.com/products/shoes.html', 'secure' => false)
 
+Custom ports
+^^^^^^^^^^^^^^^^^^^^
```
The `^` symbol count must match above line length in symbols. The PhpStorm reports this. I wonder why this haven't failed documentation build.

aik099 commented on docs/getting-started.rst:
```diff
@@ -54,6 +54,8 @@ The following configuration options are available:
 
 * ``base_url`` - allows to specify Base URL to be used to transform all relative urls from :ref:`@page-url annotations <page-url>` into absolute urls.
 
+.. note:: It is also possible to use a specific port like ``8080`` via ``http://example.com:8080``.
```
We have a lot more options, then `base_url`, but they are missing in docs current. When we added then, then `note` you've added can't be easily matched to `base_url` setting. Maybe add this in 2nd sentence after `base_url` option is described.

Text like this sounds better to me (change as you see fit):
```
If port is specified as part of `base_url` then it will be used in every built url unless specified exactly in the `@page-url` annotation.
```

evangelion1204 replied:
True, then I can drop the note itself, most people should be familiar specifying a port in a url but should still keep the custom port in the annotation, although its basic knowledge.

evangelion1204 commented:
Took you sentence and removed the info, I like it more and tweaked the text in custom ports of `@find-by`.

aik099 commented on tests/QATools/QATools/PageObject/Url/ParserTest.php:
```diff
@@ -68,6 +70,7 @@ public function testGetComponentDefault()
 		$this->assertEquals('default_path', $url_parser->getComponent('path', 'default_path'));
 		$this->assertEquals('default_query', $url_parser->getComponent('query', 'default_query'));
 		$this->assertEquals('default_fragment', $url_parser->getComponent('fragment', 'default_fragment'));
+		$this->assertEquals(80, $url_parser->getComponent('port', 80));
```
Better change port to something non-default here to indicate that what's given in 2nd parameter to `getComponent` is returned.

aik099 commented:
You're almost there. Just one more change.

evangelion1204 commented:
Ok pushed with port `8080` instead of `80`

aik099 commented:
Good work. Merging. Thanks.
